src/V3Timing.cpp

```diff
@@ -0,0 +1,1190 @@
+        setSuspendableProcess();
+        iterateChildren(nodep);
+    }
+    virtual void visit(AstNodeAssign* nodep) override {
+        if (nodep->timingControlp()) setSuspendableProcess();
+        iterateChildren(nodep);
+    }
+    virtual void visit(AstEventControl* nodep) override {
+        setDynamicProcess();
+        iterateChildren(nodep);
+    }
+    virtual void visit(AstWait* nodep) override {
+        setDynamicProcess();
+        iterateChildren(nodep);
+    }
+    virtual void visit(AstFork* nodep) override {
+        if (!nodep->joinType().joinNone()) setDynamicProcess();
+        VL_RESTORER(m_proc);
+        VL_RESTORER(m_underFork);
+        m_proc = nodep;
+        m_underFork = true;
+        setDynamicProcess();
+        iterateChildren(nodep);
+    }
+    virtual void visit(AstBegin* nodep) override {
+        VL_RESTORER(m_underFork);
+        m_underFork = false;
+        iterateChildren(nodep);
+    }
+    virtual void visit(AstNodeCCall* nodep) override {
+        if (nodep->funcp()->isCoroutine()) {
+            if (isDynamic(nodep))
+                setDynamicProcess();
+            else
+                setSuspendableProcess();
+        }
+        iterateChildren(nodep);
+    }
+    virtual void visit(AstVarRef* nodep) override {
+        if (m_proc && nodep->access().isWriteOrRW()) {
+            auto* const varp = nodep->varp();
+            varp->isWrittenBySuspendable(varp->isWrittenBySuspendable() || isSuspendable(m_proc));
+            if (!isWrittenByDynamic(varp) && isDynamicProcess()) {
+                m_repeat = true;
+                setWrittenByDynamic(varp);
+            }
+        }
+    }
+    virtual void visit(AstAssignDly* nodep) override {
+        if (!isSuspendable(m_proc)) return;
+        auto* const fl = nodep->fileline();
+        auto* const eventp = getCreateDlyEvent();
+        auto* const assignp
+            = new AstAssign{fl, nodep->lhsp()->unlinkFrBack(), nodep->rhsp()->unlinkFrBack()};
+        auto* const eventControlp = new AstEventControl{
+            fl,
+            new AstSenTree{fl, new AstSenItem{fl, VEdgeType::ET_ANYEDGE,
+                                              new AstVarRef{fl, eventp, VAccess::READ}}},
+            assignp};
+        if (m_underFork) {
+            nodep->replaceWith(eventControlp);  // No need to create another fork
+        } else {
+            auto* const forkp = new AstFork{nodep->fileline(), "", eventControlp};
+            forkp->joinType(VJoinType::JOIN_NONE);
+            nodep->replaceWith(forkp);
+        }
+        VL_DO_DANGLING(delete nodep, nodep);
+    }
+
+    //--------------------
+    virtual void visit(AstNode* nodep) override { iterateChildren(nodep); }
+
+public:
+    AstVarScope* getDlyEvent() { return m_dlyEvent; }
+
+    // CONSTRUCTORS
+    explicit TimingTransformVisitor(AstNetlist* nodep) {
+        do {
+            m_repeat = false;
+            iterate(nodep);
+        } while (m_repeat);
```

Comment:
Does it ever run more than twice (I hope not), so add an assertion the loop doesn't do more than twice.